share/p2p/peers/metrics.go

```diff
@@ -0,0 +1,275 @@
+	}
+
+	err = meter.RegisterCallback(
+		[]instrument.Asynchronous{
+			shrexPools,
+			fullNodesPool,
+			blacklisted,
+		},
+		func(ctx context.Context) {
+			for poolStatus, count := range manager.shrexPools() {
+				shrexPools.Observe(ctx, count,
+					attribute.String(poolStatusKey, string(poolStatus)))
+			}
+
+			fullNodesPool.Observe(ctx, int64(manager.fullNodes.len()),
+				attribute.String(peerStatusKey, string(peerStatusActive)))
+			fullNodesPool.Observe(ctx, int64(manager.fullNodes.cooldown.len()),
+				attribute.String(peerStatusKey, string(peerStatusCooldown)))
+
+			metrics.blacklistedPeersByReason.Range(func(key, value any) bool {
+				reason := key.(blacklistPeerReason)
+				amount := value.(int)
+				blacklisted.Observe(ctx, int64(amount),
+					attribute.String(blacklistPeerReasonKey, string(reason)))
+				return true
+			})
+		},
+	)
+
+	if err != nil {
+		return nil, fmt.Errorf("regestering metrics callback: %w", err)
+	}
+	return metrics, nil
+}
+
+func (m *metrics) observeGetPeer(source peerSource, poolSize int, waitTime time.Duration) {
+	if m == nil {
+		return
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), observeTimeout)
+	defer cancel()
+	m.getPeer.Add(ctx, 1,
+		attribute.String(sourceKey, string(source)),
+		attribute.Bool(isInstantKey, waitTime == 0))
+	m.getPeerPoolSizeHistogram.Record(ctx, int64(poolSize),
+		attribute.String(sourceKey, string(source)))
+
+	// record wait time only for async gets
+	if waitTime > 0 {
+		m.getPeerWaitTimeHistogram.Record(ctx, waitTime.Milliseconds(),
+			attribute.String(sourceKey, string(source)))
+	}
+}
+
+func (m *metrics) observeDoneResult(source peerSource, result result) {
+	if m == nil {
+		return
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), observeTimeout)
+	defer cancel()
+
+	m.doneResult.Add(ctx, 1,
+		attribute.String(sourceKey, string(source)),
+		attribute.String(doneResultKey, string(result)))
+}
+
+// validationObserver is a middleware that observes validation results as metrics
+func (m *metrics) validationObserver(validator shrexsub.ValidatorFn) shrexsub.ValidatorFn {
+	if m == nil {
+		return validator
+	}
+	return func(ctx context.Context, id peer.ID, n shrexsub.Notification) pubsub.ValidationResult {
+		res := validator(ctx, id, n)
+
+		var resStr string
+		switch res {
+		case pubsub.ValidationAccept:
+			resStr = validationAccept
+		case pubsub.ValidationReject:
+			resStr = validationReject
+		case pubsub.ValidationIgnore:
+			resStr = validationIgnore
+		default:
+			resStr = "unknown"
+		}
+
+		observeCtx, cancel := context.WithTimeout(context.Background(), observeTimeout)
+		defer cancel()
+
+		m.validationResult.Add(observeCtx, 1,
+			attribute.String(validationResultKey, resStr))
+		return res
+	}
+}
+
+// observeBlacklistPeers stores amount of blacklisted peers by reason
+func (m *metrics) observeBlacklistPeers(reason blacklistPeerReason, amount int) {
+	if m == nil {
+		return
+	}
+	for {
```

Comment:
is there some sort of cancellation for this?

Reply:
it will return on successful update

Reply:
why does it have to be in a loop?

Reply:
If the value is updated after read and before CAS, it will not get updated. This is concurrency CAS protection

Reply:
Alright.